adapter/internal/ingress/client_manager.go

```diff
@@ -64,42 +65,48 @@ func NewClientManager(connector ConnectionBuilder, connCount int, ttl time.Durat
 func (c *ClientManager) Next() v2.EgressClient {
 	for {
 		idx := atomic.AddUint64(&c.nextIdx, 1)
-		conn := (*connection)(atomic.LoadPointer(&c.connections[idx%uint64(len(c.connections))]))
-		if conn.client != nil {
+		actualIdx := int(idx % uint64(len(c.connections)))
+
+		c.mu.RLock()
+		conn := c.connections[actualIdx]
+		if conn != nil && conn.client != nil {
 			return conn.client
 		}
+		c.mu.RUnlock()
 
+		c.openNewConnection(actualIdx)
 		time.Sleep(c.retryWait)
 	}
 }
 
 func (c *ClientManager) monitorConnectionsForRolling() {
 	for range time.Tick(c.connectionTTL) {
 		for i := 0; i < len(c.connections); i++ {
-			conn := (*connection)(atomic.LoadPointer(&c.connections[i]))
-
-			if conn.closer != nil {
+			c.mu.Lock()
+			conn := c.connections[i]
+			if conn != nil && conn.closer != nil {
 				conn.closer.Close()
 			}
+			c.mu.Unlock()
 
 			c.openNewConnection(i)
 		}
 	}
 }
 
 func (c *ClientManager) openNewConnection(idx int) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
```

Comment:
We should not take out a lock while establishing a connection. While the write lock is held you can not take out a read lock which, according to @apoydence, will crush throughput.

You can establish the connection, then take out the write lock to insert into the `[]*connection`